### PR TITLE
Merge 013_FEAT_CRUD-REVIEW and 014_FEAT_CRUD_COMMENT into dev

### PR DESCRIPTION
**Pull request by BZJohnCen**
Tickets #013 and #014 were implemented together

### CODE REVIEW

File: api-handler.js [hunk 1]
```diff
@@ -1,67 +1,127 @@
...
+  let payload = event.body
+  try {
+    return await ReviewsHelper.updateComment(commentId, payload)
+  } catch (err) {
+    console.error('caught error:', err.message)
+    return Status.createErrorResponse(400, err.message)
+  }
+}
+
+// module.exports.addCommentToReview = async (event) => {
```
Comment (Praiyon) [inline api-handler.js]: delete this?

File: helpers/reviews.js [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+        console.log('New Rating Created:\n', result)
+        return newRating._id
+    } catch (err) {
+        console.error('caught err while trying to save Rating to db:\n', err.message)
+    }
+}
+
+//Returns Company obj 
+const findCompanyByName = async (eventBody) => {
+    // let newCompany = Company({
```
Comment (Praiyon) [inline helpers/reviews.js]: I think you forgot to remove this comment

File: helpers/reviews.js [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+    // })
+    try { 
+        let foundCompany = await db(MONGO_URL, () => {
+            return Company.find({ name: eventBody.company_name.toLowerCase().trim() })
+        })
+        console.log('Company Found:\n', foundCompany)
+        if (foundCompany.length > 0) {
+            foundCompany = foundCompany[0]
+        } else {
+            // foundCompany = newCompany
```
Comment (Praiyon) [inline helpers/reviews.js]: remove this as well

File: helpers/reviews.js [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+        user: foundUserId,
+        company: foundCompany._id,
+        upvotes: [],
+        downvotes: [],
+        comments: []
+    })
+    try {
+        let result = await db(MONGO_URL, () => {
+            return newReview.save().catch(err => {
+                console.error('caught err when trying to save to db:\n', err.message)
```
Comment (Praiyon) [inline helpers/reviews.js]: this looks like a redundant catch. The catch block outside of the try statement does the same thing

File: helpers/reviews.js [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+}
+    //deleteReview 3.3
+module.exports.deleteReview = async (reviewId) => {
+    try {
+        let result = await db(MONGO_URL, () => {
+            return Review.findOneAndDelete({
+                _id: reviewId
+            })
+        })
+        console.log('Deleted Review obj:\n', result)
```
Comment (Praiyon) [inline helpers/reviews.js]: you should log this after `if (result)`

File: helpers/reviews.js [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+            return Status.createSuccessResponse(204, {
+                message: "Comment successfully UPDATED."
+            })
+    } catch (err) {
+        console.error('update comment caught error:', err.message)
+        return Status.createErrorResponse(400, err.message)
+    }
+}
+
+// module.exports.addCommentToReview = async (payload) => {
```
Comment (Praiyon) [inline helpers/reviews.js]: forgot to delete?

File: helpers/reviews.js [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+}
+
+//--------------- EXPORTED FUNCTIONS ---------------
+
+//013_FEAT_CRUD-REVIEW
+    //createReview 1.0
+module.exports.createReview = async (payload) => {
+    console.log('payload:\n', payload)
+    let foundUserId = await findUserId(payload) //find user
+    console.log('foundUserId:\n', foundUserId)
```
Comment (Praiyon) [inline helpers/reviews.js]: The same console log is made inside the `findUserId` function.

File: helpers/reviews.js [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+    let newRatingId = await createRatingAndSave(payload) //create Rating & add to db
+    console.log('newRatingId:\n', newRatingId)
```
Comment (Praiyon) [inline helpers/reviews.js]: same as above

**Review by Praiyon — CHANGES_REQUESTED** [review]
Just some small stuff

**Review by BZJohnCen — COMMENTED** [review]
PR suggestions changes made

File: helpers/reviews.js [hunk 9]
```diff
@@ -16,3 +23,337 @@ module.exports.getFlaggedReviews = () => {
...
+        console.error('review does not exist:\n', err.message)
+    }
+}
+
+const addCommentToReview = async (reviewId, commentId) => {
+    try {
+        //grab existing comments from review obj
+        let review = await getReviewById(reviewId)
+        let existingComments = review.comments
+        console.log("commentsList count:", existingComments.length)
```
Comment (byan1197) [inline helpers/reviews.js]: idk if we need this log

File: api-handler.js [hunk 10]
```diff
@@ -1,70 +1,121 @@
...
-//Returns an error response
-const sendErrorResponse = (statusCode, errorMessage) => {
-  console.error('sendErrorRepsonse: console logging error msg:\n', errorMessage)
-  return {
-    statusCode: statusCode,
-    headers: { "Content-Type": "application/json", "Access-Control-Allow-Origin": "*" },
-    body: JSON.stringify({ error: errorMessage })
+  //updateReview 2.1
+module.exports.updateReviewFields = async (event) => {
+  let payload = event.body
```
Comment (byan1197) [inline api-handler.js]: JSON parse this

File: api-handler.js [hunk 11]
```diff
@@ -1,70 +1,121 @@
...
-
-//Returns an error response
-const sendErrorResponse = (statusCode, errorMessage) => {
-  console.error('sendErrorRepsonse: console logging error msg:\n', errorMessage)
-  return {
-    statusCode: statusCode,
-    headers: { "Content-Type": "application/json", "Access-Control-Allow-Origin": "*" },
-    body: JSON.stringify({ error: errorMessage })
+  //updateReview 2.1
+module.exports.updateReviewFields = async (event) => {
```
Comment (byan1197) [inline api-handler.js]: change name updateReview

File: api-handler.js [hunk 12]
```diff
@@ -1,70 +1,121 @@
...
+  let companyId = event.pathParameters.company_id
+  try {
+    return await ReviewsHelper.updateReviewCompany(companyId, payload)
+  } catch (err) {
+      console.error('company does not exist:\n', err.message)
+      return Status.createErrorResponse(400, err.message)
+  }
+}
+  //deleteReview 3.1
+module.exports.deleteRating = async (event) => {
```
Comment (byan1197) [inline api-handler.js]: demote to a helper function

File: api-handler.js [hunk 13]
```diff
@@ -1,70 +1,121 @@
...
+  let ratingId = event.pathParameters.rating_id
+  try {
+    return await ReviewsHelper.deleteRating(ratingId)
+  } catch (err) {
+    console.error('caught error:', err.message)
+    return Status.createErrorResponse(400, err.message)
+  }
+}
+  //deleteReview 3.2
+module.exports.deleteAllComments = async (event) => {
```
Comment (byan1197) [inline api-handler.js]: demote

File: api-handler.js [hunk 14]
```diff
@@ -1,70 +1,121 @@
...
+  let payload = event.body //pass in comments array
+  try {
+    return await ReviewsHelper.deleteAllComments(payload)
+  } catch (err) {
+    console.error('caught error:', err.message)
+    return Status.createErrorResponse(400, err.message)
+  }
+}
+  //deleteReview 3.3
+module.exports.deleteReview = async (event) => {
```
Comment (byan1197) [inline api-handler.js:53]: have the helper functions

File: api-handler.js [hunk 15]
```diff
@@ -1,70 +1,121 @@
...
+    return await ReviewsHelper.createComment(payload)
   } catch (err) {
     console.error('caught error:', err.message)
-    return sendErrorResponse(400, err.message)
+    return Status.createErrorResponse(400, err.message)
   }
-};
+}
+  //deleteComment
+module.exports.deleteComment = async (event) => {
```
Comment (byan1197) [inline api-handler.js:76]: patch request to remove content and user, but keep the object

File: helpers/reviews.js [hunk 16]
```diff
@@ -1,18 +1,354 @@
...
+        })
+        console.log('foundUser:\n', foundUser)
+        return foundUser[0]._id
+    } catch (err) {
+        console.error('user does not exist:\n', err.message)
+    }
+}
+
+//Creates a new Rating obj and saves to db. Returns rating ID
+const createRatingAndSave = async (eventBody) => {
```
Comment (byan1197) [inline helpers/reviews.js]: need its own lambda

File: helpers/reviews.js [hunk 17]
```diff
@@ -1,18 +1,354 @@
 const Review = require('@pbnj-xintern/xintern-commons/models/Review')
```
Comment (byan1197) [inline helpers/reviews.js:1]: use the commons payload checker

**Review by byan1197 — CHANGES_REQUESTED** [review]

File: helpers/reviews.js (hunk 16 above)
Comment (BZJohnCen) [reply helpers/reviews.js]: renamed to `createRating` and remains a helper function

File: helpers/reviews.js (hunk 9 above)
Comment (BZJohnCen) [reply helpers/reviews.js]: removed logs

File: helpers/reviews.js [hunk 18]
```diff
@@ -1,18 +1,350 @@
...
+
+//014_FEAT_CRUD_COMMENT
+    //createComment + link to Review
+module.exports.createComment = async (reviewId, payload) => {
+    let newComment = Comment({
+        _id: new mongoose.Types.ObjectId(),
+        content: payload.content,
+        upvotes: [],
+        downvotes: [],
+        parentComment: (Boolean(payload.parent_comment_id)) ? payload.parent_comment_id : null
```
Comment (Praiyon) [inline helpers/reviews.js]: I dont think this `Boolean` is necessary

**Review by BZJohnCen — COMMENTED** [review]
existing functions and endpoints updated according to doc

File: helpers/reviews.js [hunk 19]
```diff
@@ -1,18 +1,361 @@
...
+//--------------- EXPORTED FUNCTIONS ---------------
+
+//013_FEAT_CRUD-REVIEW
+    //createReview 1.0
+module.exports.createReview = async (payload) => {
+    console.log('payload:\n', payload)
+    let foundUserId = await findUserId(payload)
+    let ratingId = await createRating(payload)
+    let foundCompany = await findCompanyByName(payload)
+    console.log('foundCompany:\n', foundCompany)
```
Comment (byan1197) [inline helpers/reviews.js]: ideally we would need to check if foundUserId, ratingId, and foundCompany are not null

File: helpers/reviews.js [hunk 20]
```diff
@@ -1,18 +1,361 @@
...
+        parentComment: (payload.parent_comment_id) ? payload.parent_comment_id : null
+    })
+    try {
+        let result = await db(MONGO_URL, () => {
+            return newComment.save().catch(err => {
+                console.error('caught err when trying to save to db:\n', err.message)
+            })
+        })
+        console.log('new comment:\n', result)
+        let newCommentId = result._id
```
Comment (byan1197) [inline helpers/reviews.js:302]: null error handling

File: helpers/reviews.js [hunk 21]
```diff
@@ -1,18 +1,361 @@
...
+    //deleteComment - patch request to remove content and user, but keep the object
+module.exports.deleteComment = async (commentId) => {
+    try {
+        let result = await db(MONGO_URL, () => {
+            return Comment.findOneAndUpdate(commentId, {
+                author: null, //handle err msg client side
+                content: "[this comment has been removed.]"
+            })
+        })
+        if (result)
```
Comment (byan1197) [inline helpers/reviews.js]: what happens if no result?

**Review by byan1197 — APPROVED** [review]

**Pull request merged**